synapse/handlers/room_member.py

```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
                 return data["mxid"]
 
         except IOError as e:
-            logger.warn("Error from identity server lookup: %s" % (e,))
+            logger.warn("Error from v1 identity server lookup: %s" % (e,))
+
+        return None
+
+    @defer.inlineCallbacks
+    def _lookup_3pid_v2(
+        self, id_server, id_access_token, medium, address, hash_details
+    ):
+        """Looks up a 3pid in the passed identity server using v2 lookup.
+
+        Args:
+            id_server (str): The server name (including port, if required)
+                of the identity server to use.
+            id_access_token (str): The access token to authenticate to the identity server with
+            medium (str): The type of the third party identifier (e.g. "email").
+            address (str): The third party identifier (e.g. "foo@example.com").
+            hash_details (dict[str, str|list]): A dictionary containing hashing information
+                provided by an identity server.
+
+        Returns:
+            Deferred[str|None]: the matrix ID of the 3pid, or None if it is not recognised.
+        """
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details.get("algorithms")
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if not supported_lookup_algorithms or not lookup_pepper:
+            raise SynapseError(
+                500, "Invalid hash details received from identity server: %s, %s"
+            )
+
+        # Check if any of the supported lookup algorithms are present
+        if LookupAlgorithm.SHA256 in supported_lookup_algorithms:
+            # Perform a hashed lookup
+            lookup_algorithm = LookupAlgorithm.SHA256
+
+            # Hash address, medium and the pepper with sha256
+            to_hash = "%s %s %s" % (address, medium, lookup_pepper)
+            lookup_value = sha256_and_url_safe_base64(to_hash)
+
+        elif LookupAlgorithm.NONE in supported_lookup_algorithms:
+            # Perform a non-hashed lookup
+            lookup_algorithm = LookupAlgorithm.NONE
+
+            # Combine together plaintext address and medium
+            lookup_value = "%s %s" % (address, medium)
+
+        else:
+            logger.warn(
+                "None of the provided lookup algorithms of %s are supported: %s",
+                id_server,
+                supported_lookup_algorithms,
+            )
+            raise SynapseError(
+                400,
```

Comment:
is this specced behaviour? (as opposed to returning None)? if so, can you document this behaviour on the docstring here and on _lookup_3pid?

Reply:
MSC2134 only mentions an `M_INVALID_PARAM` on attempting to use a lookup algorithm that the identity server doesn't support, but doesn't cover the client not supporting any of the algorithms the identity server does.

I think we just assumed lookup could not happen in this case.

So it might just be easier to have lookup "fail" by returning no results. The client might be a bit confused about why the 3PID lookup failed though.

Reply:
perhaps you could ask the MSC author to clarify the expected behaviour?

> I think we just assumed lookup could not happen in this case.

I'm not quite sure what you mean by this.

Reply:
> perhaps you could ask the MSC author to clarify the expected behaviour?

That'd be me.

> > I think we just assumed lookup could not happen in this case.
> 
> I'm not quite sure what you mean by this.

You would need to pick a different identity server.

Reply:
After discussing in the privacy room we're thinking a `500` error since old clients will not know what to do with any special error messages.

Reply:
And now 400 again.